modules/rgbd/src/opencl/hash_tsdf.cl

```diff
@@ -0,0 +1,702 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+// This code is also subject to the license terms in the LICENSE_KinectFusion.md file found in this module's directory
+
+#define NAN_ELEMENT -2147483647
+#define USE_INTERPOLATION_IN_GETNORMAL 1
+
+typedef __INT8_TYPE__ int8_t;
+typedef __INT32_TYPE__ int32_t;
+
+typedef int8_t TsdfType;
+typedef uchar WeightType;
+
+struct TsdfVoxel
+{
+    TsdfType tsdf;
+    WeightType weight;
+};
+
+struct Volume_NODE
+{
+    int4 idx;
+    int32_t row;
+    int32_t nextVolumeRow;
+    int32_t dummy;
+    int32_t dummy2;
+};
+
+static inline TsdfType floatToTsdf(float num)
+{
+    int8_t res = (int8_t) ( (num * (-128)) );
+    res = res ? res : (num < 0 ? 1 : -1);
+    return res;
+}
+
+static inline float tsdfToFloat(TsdfType num)
+{
+    return ( (float) num ) / (-128);
+}
+
+__kernel void preCalculationPixNorm (__global char * pixNormsPtr,
+                                     int pixNormsStep, int pixNormsOffset,
+                                     int pixNormsRows, int pixNormsCols,
+                                     const __global float * xx,
+                                     const __global float * yy)
+{
+    int i = get_global_id(0);
+    int j = get_global_id(1);
+    if (i < pixNormsRows && j < pixNormsCols)
+    {
+        *(__global float*)(pixNormsPtr + pixNormsOffset + i*pixNormsStep + j*sizeof(float)) = sqrt(xx[j] * xx[j] + yy[i] * yy[i] + 1.0f);
+    }
+}
+
+static uint calc_hash(int3 x)
+{
+    unsigned int seed = 0;
+    unsigned int GOLDEN_RATIO = 0x9e3779b9;
+    seed ^= x.s0 + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    seed ^= x.s1 + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    seed ^= x.s2 + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    return seed;
+}
+
+
+static int findRow(__global const struct Volume_NODE * hash_table, int3 indx,
+                   int list_size, int bufferNums, int hash_divisor)
+{
+    int bufferNum = 0;
+    int hash = calc_hash(indx) % hash_divisor;
+    int i = (bufferNum * hash_divisor + hash) * list_size;
+    while (i >= 0)
+    {
+        struct Volume_NODE v = hash_table[i];
+        if (all(v.idx.s012 == indx.s012))
+            return v.row;
+        else
+            i = v.nextVolumeRow;
+    }
+
+    return -1;
+}
+
+//TODO: make hashDivisor a power of 2
+//TODO: put it to this .cl file as a constant
```

Comment:
TODO: implement these TODOs:
* create a constant `globalHashSetDivisor` equal to the one in CustomHashSet on host
* remove `hashDivisor` argument from all functions
* do not pass it from host side either